src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketGitSCMBuilder.java

```diff
@@ -0,0 +1,272 @@
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.Util;
+import hudson.plugins.git.GitSCM;
+import hudson.plugins.git.browser.BitbucketWeb;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import jenkins.plugins.git.AbstractGitSCMSource;
+import jenkins.plugins.git.GitSCMBuilder;
+import jenkins.scm.api.SCMHead;
+import jenkins.scm.api.SCMRevision;
+import jenkins.scm.api.SCMSource;
+import jenkins.scm.api.mixin.ChangeRequestCheckoutStrategy;
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * A {@link GitSCMBuilder} specialized for bitbucket.
+ *
+ * @since 2.2.0
+ */
+public class BitbucketGitSCMBuilder extends GitSCMBuilder<BitbucketGitSCMBuilder> {
+
+    /**
+     * The {@link BitbucketSCMSource} who's {@link BitbucketSCMSource#getOwner()} can be used as the context for
+     * resolving credentials.
+     */
+    @NonNull
+    private final BitbucketSCMSource scmSource;
+
+    /**
+     * The clone links for cloning the source repository and origin pull requests (but links will need tweaks for
+     * fork pull requests)
+     */
+    @NonNull
+    private List<BitbucketHref> cloneLinks = Collections.emptyList();
+
+    /**
+     * Constructor.
+     *
+     * @param scmSource     the {@link BitbucketSCMSource}.
+     * @param head          the {@link SCMHead}
+     * @param revision      the (optional) {@link SCMRevision}
+     * @param credentialsId The {@link IdCredentials#getId()} of the {@link Credentials} to use when connecting to
+     *                      the {@link #remote()} or {@code null} to let the git client choose between providing its own
+     *                      credentials or connecting anonymously.
+     */
+    public BitbucketGitSCMBuilder(@NonNull BitbucketSCMSource scmSource, @NonNull SCMHead head,
+                                  @CheckForNull SCMRevision revision, @CheckForNull String credentialsId) {
+        // we provide a dummy repository URL to the super constructor and then fix is afterwards once we have
+        // the clone links
+        super(head, revision, /*dummy value*/scmSource.getServerUrl(), credentialsId);
+        withoutRefSpecs();
+        if (head instanceof PullRequestSCMHead) {
+            if (scmSource.buildBitbucketClient() instanceof BitbucketCloudApiClient) {
+                // TODO fix once Bitbucket Cloud has a fix for https://bitbucket.org/site/master/issues/5814
+                String branchName = ((PullRequestSCMHead) head).getBranchName();
+                withRefSpec("+refs/heads/" + branchName + ":refs/remotes/@{remote}/" + head.getName());
+            } else {
+                String pullId = ((PullRequestSCMHead) head).getId();
+                withRefSpec("+refs/pull-requests/" + pullId + "/from:refs/remotes/@{remote}/" + head.getName());
```

Comment:
Something having to be installed on bitbucket server is not a nice solution.
Currently one of our server is managed in a company level shared environment and it's difficult to install plugins; probably it makes sens to no install a lot of hook plugins because each tool want their own.
Relying on pulling data is the desired solution in my case.